### PR TITLE
Verifier abstract base

### PR DESCRIPTION
**Pull request by EdNoepel**
Moved common verifier functionality from `perennial-verifier` package and improved documentation.

### CODE REVIEW

File: contracts/verifier/interfaces/IVerifierBase.sol [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+    /// @notice Cancels a group nonce
+    /// @param group The group nonce to cancel
+    function cancelGroup(uint256 group) external;
+
+    /// @notice Cancels a group for an account via a signed message
+    /// @param groupCancellation The group cancellation message
+    /// @param signature The signature of the account for the group cancellation
+    function cancelGroupWithSignature(GroupCancellation calldata groupCancellation, bytes calldata signature) external;
+}
```
Comment (kbrizzle) [inline contracts/verifier/interfaces/IVerifierBase.sol:64]: nit: got a few of these missing last lines

File: contracts/verifier/test/VerifierBaseTester.sol [hunk 2]
```diff
@@ -0,0 +1,10 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity ^0.8.13;
+
+import { EIP712 } from "@openzeppelin/contracts/utils/cryptography/EIP712.sol";
+import { VerifierBase } from "../VerifierBase.sol";
+
+/// @dev Empty implementation for the sole purpose of testing base class
+contract VerifierBaseTester is VerifierBase {
+    constructor() EIP712("Perennial Root Unit Tests", "1.0.0") { }
```
Comment (kbrizzle) [inline contracts/verifier/test/VerifierBaseTester.sol:9]: nit: this library is not actually a Perennial specific thing -- we use it across a few different projects.

Comment (EdNoepel) [reply contracts/verifier/test/VerifierBaseTester.sol]: resolved in c8a7201b2c82bcba237def531028fd902c4c99eb

File: contracts/verifier/VerifierBase.sol [hunk 3]
```diff
@@ -0,0 +1,85 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity ^0.8.13;
+
+import { EIP712 } from "@openzeppelin/contracts/utils/cryptography/EIP712.sol";
+import { ECDSA } from "@openzeppelin/contracts/utils/cryptography/ECDSA.sol";
+import { Common, CommonLib } from "./types/Common.sol";
+import { GroupCancellation, GroupCancellationLib } from "./types/GroupCancellation.sol";
+import "./interfaces/IVerifierBase.sol";
```
[the remark this answers is not in the document]
Comment (EdNoepel) [reply contracts/verifier/VerifierBase.sol]: resolved in c8a7201b2c82bcba237def531028fd902c4c99eb

File: README.md [hunk 4]
```diff
@@ -2,6 +2,17 @@
 
 Core library for DeFi.
 
+## Facilities
```
[the remark this answers is not in the document]
Comment (EdNoepel) [reply README.md:5]: I'm missing a few that I'm unfamiliar with, but figured better having incomplete descriptions than none at all.  I would also like to build out documentation for _attribute_ some day.

**Review by kbrizzle — APPROVED** [review]

**Review by arjun-io — APPROVED** [review]

**Pull request merged**